articles/151_roslaunch_xml.md

```diff
@@ -0,0 +1,744 @@
+</group>
+```
+
+#### <arg> Tag
+
+##### Description
+
+The <arg> tag allows for launch file configuration via the command-line or when including it via an <include> tag.
+Arguments are limited to the scope of their definition and thus, have to be explictly passed to included files if any.
+
+##### Examples
+
+```xml
+<arg name="publish_frequency" default="10"/>
+<arg name="output_path" doc="Output path for some processing pipeline"/>
+```
+
+#### <node> Tag
+
+##### Description
+
+The <node> tag allows for executing a ROS node in the form of local OS process.
+
+##### Examples
+
+```xml
+<node package="ros_demos" executable="publisher">
+  <param name="publish_frequency">10</param>
+  <remap from="generic_topic_name" to="my_topic"/>
+</node>
+```
+
+#### <param> Tag
+
+##### Description
+
+The <param> tag allows for setting a ROS parameter of a ROS node.
+
+##### Examples
+
+```xml
+<param name="some_numeric_param" value="100.2"/>
+<param name="some_list_param" value="Some phrase,100.0,true" sep=","/>
+```
+
+#### <params> Tag
+
+##### Description
+
+The <params> tag allows to either bring ROS parameters from a YAML parameters file or to nest a <param> definitions under an appropriate name (i.e. to make a map of them).
+
+##### Examples
+
+```xml
+<params from="path/to/param/file.yml"/>
+<params ns="some_param_group">
+  <param name="some_integer_param" value="10"/>
+</params>
+```
+
+#### <remap> Tag
+
+##### Description
+
+The <remap> tag allows for ROS name remapping of a <node> instance.
+
+##### Examples
+
+```xml
+<remap from="chatter" to="my_chatter"/>
+```
+
+#### <env> Tag
+
+##### Description
+
+The <env> tag allows for modifying a <node>'s OS process environment.
+
+##### Examples
+
+```xml
+<env name="RMW_IMPLEMENTATION" value="rmw_fastrtps_cpp"/>
+```
+
+## Dynamic Configuration
+
+### Substitution Syntax
+
+All substitutions are enclosed by `$(...)`.
+
+### Built-in Substitutions
+
+`$(find-pkg pkg-name)`
+: Substituted by the path to package installation directory in the local filesystem.
+  Forward and backwards slashes will be resolved to the local filesystem convention.
+  Substitution will fail if the package cannot be found.
+
+`$(find-exec exec-name)`
+: Substituted by the path to the executable in the local filesystem.
+  Lookups make use of the PATH environment variable.
+  Forward and backwards slashes will be resolved to the local filesystem convention.
+  Substitution will fail if the executable cannot be found.
+
+`$(arg name)`
+: Substituted by the value of the named argument. 
+  Substitution will fail if the named argument does not exist.
+
+`$(env env-var [default-value])`
+: Substituted by the value of the given environment variable
+  Substitution will fail if the variable is not set, unless default values are provided.
+
+`$(dirname)`
+: Substituted by the current launch file directory name.
+  Substitution will always succeed.
+
+#### User-defined Substitutions
+
+TBD
```

Comment:
This may be covered by the launch configuration concept, which you can set with an action and get with a substitution:

https://github.com/ros2/launch/blob/master/launch/launch/actions/set_launch_configuration.py
https://github.com/ros2/launch/blob/master/launch/launch/substitutions/launch_configuration.py

Reply:
I was thinking of user defined extensions, as in `Substitution` subclasses. 

But still, yours is a valid point. I think we can add a `<let var="" value=""/>` tag for that action and use it with `$(var name)`, just like arguments. PTAL.